frame/support/src/traits/hooks.rs

```diff
@@ -507,6 +526,10 @@ mod tests {
 			<(Foo, (Bar, Baz))>::try_on_runtime_upgrade(true).unwrap();
 			assert_eq!(Pre::take(), vec!["Foo", "Bar", "Baz"]);
 			assert_eq!(Post::take(), vec!["Foo", "Bar", "Baz"]);
+
+			// calling pre_upgrade and post_upgrade directly on tuple of pallets fails
+			assert!(<(Foo, (Bar, Baz))>::pre_upgrade().is_err());
+			assert!(<(Foo, (Bar, Baz))>::post_upgrade(vec![]).is_err());
```

Comment:
Then why do pallets still implement `OnRuntimeUpgrade`, if its not working as intended?

Reply:
what if we implement `pre/post_upgrade` for `Tuple` since they are public and leave the note in trait doc that they should not be used with `try_on_runtime_upgrade`, since it already call's them

Reply:
I don't think a trait doc comment is sufficient, most people would never see it because they never touch `Executive` themselves.

Imo it could be even worse because the bug is even more subtle: pre/post still execute but not in the way that is expected. 

This is why I made it error, so there is no chance that a runtime accidentally has them misconfigured.

Reply:
yes, the current solution also looks good to me.
but it just looks to me that the problem is the misuse of the trait.
if someone calls `pre/post_upgrade` around the default impl of `try_on_runtime_upgrade` it is also wrong.

Reply:
I see what you're saying. Will update.